File: ALGORITHM/Leetcode/java/005_Longest_Palindromic_Substring.java
/**
 * @FileName :  005_Longest_Palindromic_Substring.java
 * @Project : LeetCode
 * @Date : 2020. 8. 29.
 * @author : AoN
 * @Link : https://leetcode.com/problems/longest-palindromic-substring/
 * @참고 : https://zkf85.github.io/2019/03/26/leetcode-005-longest-palindrome
 *
 */

// Approach 1 : Dynamic Programming
 class Solution {
   public String longestPalindrome(String s) {
	   int len = s.length();
	   char[] c = s.toCharArray();
	   Boolean[][] dp = new Boolean[len][len];
	   int start = 0, end = 0;

	   for (int i= len-1; i>= 0; i--) {
		   for (int j = i; j<len; j++) {
               if (j-i < 3) { //base case
                 dp[i][j] = (c[i] == c[j]);
               } else {
                  dp[i][j] = (c[i] == c[j] && dp[i+1][j-1]);
               }

			   if (dp[i][j] && (end-start <= j-i)) {
				   start = i;
				   end = j;
			   }
		   }
	   }
	   return s.substring(start, end+1);
   }
}

File: Leetcode/python/005_Longest_Palindromic_Substring.py
"""
 * @FileName : 005_Longest_Palindromic_Substring.py
 * @Project : LeetCode
 * @Date : 2020. 8. 29.
 * @author : AoN
 * @Link : https://leetcode.com/problems/longest-palindromic-substring/
 * @Description : Manacher Algorithm
 * 
"""


class Solution:
    def longestPalindrome(self, s: str) -> str:
        ss = ''
        for i in s:
            ss = ss + '#' + i
        ss = ss + '#'
        p = [0] * 2001
        start, end, maxlen, mx = 0, 0, 0, 0

        for i in range(len(ss)):
            p[i] = min(p[2 * id - i], mx - i) if i < mx else 1

            while (i - p[i]) >= 0 and i + p[i] < len(ss) and ss[i - p[i]] == ss[i + p[i]]:
                p[i] += 1

            if mx < i + p[i]:
                id = i
                mx = i + p[i]

            if maxlen < (p[i] - 1):
                maxlen = p[i] - 1
                start = i + 1 - p[i]
                end = i + p[i] - 1

        return ss[start:end + 1].replace('#', '')

s ="babad"
answer = Solution().longestPalindrome(s)